[DOCS] Apiserver improve docs readability (#3564)
ray-project/kuberay: apiserver/HACluster.md

# Creating HA cluster with API Server

> One of the issue for long-running Ray application (e.g. RayServe) is that if Ray head node

Reviewer: ```suggestion
One of the issues for long-running Ray application (e.g. RayServe) is that if the Ray head node
```

dies, the whole cluster has to be restarted. Fortunately, KubeRay cluster solved it by
introducing [Fault Tolerance Ray Cluster](https://docs.ray.io/en/master/cluster/kubernetes/user-guides/kuberay-gcs-ft.html).

> The RayCluster with high availability can also be created in API server, which aims to

Reviewer: ```suggestion
The RayCluster with high availability can also be created with the API server, which aims to
```

ensure a high availability Global Control Service (GCS) data. The GCS manages

> cluster-level metadata by storing all data in memory, which is lack of fault tolerance. A

Reviewer: ```suggestion
cluster-level metadata by storing all data in memory, which lacks fault tolerance. A
```

> single failure can cause the entire RayCluster to fail. To enable GCS's fault tolerance,

Reviewer 2: ```suggestion
single head node failure can cause the entire RayCluster to fail. To enable GCS's fault tolerance,
```

> we should have a highly available Redis so that when GCS restart, it can resume its

Reviewer: ```suggestion
we should have a highly available Redis so that when GCS restarts, it can resume its
```

> status by retrieve previous data from the Redis instance.

Reviewer: ```suggestion
status by retrieving previous data from the Redis instance.
```

We will provide a detailed example on how to create this highly available API Server.

## Setup

### Setup Ray Operator and API Server

Refer to [README](README.md) for setting up KubRay operator and API server.

## Example

Before going through the example, remove any running RayClusters to ensure a successful
run through of the example below.

```sh
kubectl delete raycluster --all
```

### Create external Redis cluster

A comprehensive documentation on creating Redis cluster on Kubernetes can be found
[here]( https://www.dragonflydb.io/guides/redis-kubernetes). For this example we will use a rather simple
[RedisYAML]. Simply download this YAML file and run:

```sh
kubectl create ns redis
kubectl apply -f redis.yaml -n redis
```

Note that we created a new `redis` namespace and deploy redis in it.

Alternatively, if you run on the cloud you can use managed version of HA Redis, which will not require
you to stand up, run, manage and monitor your own version of redis.

> Check if the redis is successfully set up with following command. You should see

Reviewer: ```suggestion
Check if the redis is successfully set up with the following command. You should see
```

`redis-config` in the list:

```sh
kubectl get configmaps -n redis

# NAME               DATA   AGE
# redis-config       1      19s
```

### Create Redis password secret

> Before creating your cluster, you need to create secret in the

Reviewer: ```suggestion
Before creating your cluster, you need to create a secret in the
```

namespace where you are planning to create your Ray cluster (remember, that secret is visible only within a given
namespace). To create a secret for using external redis, please download the [Secret] and

> run following command:

Reviewer: ```suggestion
run the following command:
```

```sh
kubectl apply -f redis_passwrd.yaml
```

### Install ConfigMap

We will use this [ConfigMap] which contains code for our example. For the real world
cases, it is recommended to pack user code in an image.

Please download the config map and deploy it with following command:

```sh
kubectl apply -f code_configmap.yaml
```

### Create RayCluster

> Use following command to create a compute template and a RayCluster:

Reviewer: ```suggestion
Use the following command to create a compute template and a RayCluster:
```

```sh
curl -X POST 'localhost:31888/apis/v1/namespaces/default/compute_templates' \
    --header 'Content-Type: application/json' \
    --data @docs/api-example/compute_template.json

curl -X POST 'localhost:31888/apis/v1/namespaces/default/clusters' \
    --header 'Content-Type: application/json' \
    --data @docs/api-example/ha_clusters.json
```

To enable the RayCluster's GCS fault tolerance feature, we added the annotation:

```json
"annotations" : {
    "ray.io/ft-enabled": "true"
}
```

For connecting to Redis, we set the following content in `rayStartParams` of

> `headGroupSpec`: We also added following content in `rayStartParams` of `headGroupSpec`,

Reviewer: ```suggestion
`headGroupSpec`: We also added the following content in `rayStartParams` of `headGroupSpec`,
```

which set the Redis password and the number of cpu. Setting `num-cpu` to 0 ensures that no
application code runs on a head node.

> ```json

Reviewer: this doesn't look like a json.

Author: Thanks! I added {} around it.

"redis-password:: "$REDIS_PASSWORD"
"num-cpu": "0"
```

Here, the `$REDIS_PASSWORD` is defined in `headGroupSpec`'s environment variable below:

```json
"environment": {
    "values": {

>         "RAY_REDIS_ADDRESS": "redis.redis.svc.cluster.local:6379"

Reviewer 2: just curious (I never used HA ray before), why `redis.redis.` instead of `redis.`?

Author: I think the first `redis` is the service name, while the second is the namespace

    },
    "valuesFrom": {
        "REDIS_PASSWORD": {
            "source": 1,
            "name": "redis-password-secret",
            "key": "password"
        }
    }
},
```

For the `workerGroupSpecs`, we set the `gcs_rpc_server_reconnect_timeout` environment
variable, which controls the GCS heartbeat timeout (default 60 seconds). This controls how
long after the head node dies do we kill the worker node. While it takes time to restart

> the head node, we want this values to be large enough to prevent the worker node being

Reviewer: ```suggestion
the head node, we want these values to be large enough to prevent the worker node being
```

killed during the restarting period.

```json
"environment": {
    "values": {
        "RAY_gcs_rpc_server_reconnect_timeout_s": "300"
    }
},
```

### Validate that RayCluster is deployed correctly

> Run following command to get list of pods running. You should see one head and worker node

Reviewer: ```suggestion
Run the following command to get a list of pods running. You should see one head and worker node
```

like below:

```sh
kubectl get pods
# NAME                                READY   STATUS    RESTARTS   AGE
# ha-cluster-head                     1/1     Running   0          2m36s
# ha-cluster-small-wg-worker-22lbx    1/1     Running   0          2m36s
```

### Create an Actor

Before we try to trigger the restoration, we need to find a way to validate our GCS restore
is working correctly. We will validate this by creating a detached actor. If it still
exists and functions after the head node deletion and restoration, we can confirm that the
GCS data is restored correctly.

> Run following command for creating a detached actor. Please change `ha-cluster-head` to

Reviewer: ```suggestion
Run the following command for creating a detached actor. Please change `ha-cluster-head` to
```

your head node's name. Note that the `detached_actor.py` file is defined in the
[ConfigMap] we installed earlier and mounted to the head node:

```sh
kubectl exec -it ha-cluster-head -- python3 /home/ray/samples/detached_actor.py
```

Then, open a new termianl and use port-forward to enable accessing to the Ray dashboard.
The dashboard can be accessed through `http://localhost:8265`:

```sh
kubectl port-forward pod/ha-cluster-head 8265:8265
```

In the dashboard, You can see 2 nodes in the Cluster pane, which is head and worker:

![hacluster-dashboard-cluster](img/hacluster-dashboard-cluster.png)

If you go to the Actor pane, you can see the actor we created earlier:

![hacluster-dashboard-actor](img/hacluster-dashboard-actor.png)

### Trigger the GCS restore

To trigger the restoration, we can simply delete the head node with:

```sh
kubectl delete pods ha-cluster-head
```

If you list the pods now, you can see a new head node is recreated

```sh
kubectl get pods
# NAME                                READY   STATUS    RESTARTS   AGE
# ha-cluster-head                     0/1     Running   0          5s
# ha-cluster-small-wg-worker-tpgqs    1/1     Running   0          9m19s
```

> Note that only head node will be recreated, while the worker node stays as is.

Reviewer: ```suggestion
Note that only the head node will be recreated, while the worker node stays as is.
```

# Creating HA cluster with API Server

One of the issue for long-running Ray application (e.g. RayServe) is that if Ray head node
dies, the whole cluster has to be restarted. Fortunately, KubeRay cluster solved it by
introducing [Fault Tolerance Ray Cluster](https://docs.ray.io/en/master/cluster/kubernetes/user-guides/kuberay-gcs-ft.html).

The RayCluster with high availability can also be created in API server, which aims to
ensure a high availability Global Control Service (GCS) data. The GCS manages
cluster-level metadata by storing all data in memory, which is lack of fault tolerance. A
single failure can cause the entire RayCluster to fail. To enable GCS's fault tolerance,

> we should have a highly available Redis so that when GCS restart, it can resume its

Reviewer 2: Just want to make sure ray's HA mode relies on redis, but not necessarily HA redis, right?

Author: Yes!

Reviewer 2: I prefer using the word "redis" instead of "HA redis" here, these are two different products from cloud vendor offering.

Reviewer 2: redis itself doesn't provide persistence as well.

long after the head node dies do we kill the worker node. While it takes time to restart
the head node, we want this values to be large enough to prevent the worker node being
killed during the restarting period.

```json
"environment": {
    "values": {
        "RAY_gcs_rpc_server_reconnect_timeout_s": "300"
    }
},
```

### Validate that RayCluster is deployed correctly

Run following command to get list of pods running. You should see one head and worker node
like below:

```sh
kubectl get pods
# NAME                                READY   STATUS    RESTARTS   AGE
# ha-cluster-head                     1/1     Running   0          2m36s
# ha-cluster-small-wg-worker-22lbx    1/1     Running   0          2m36s
```

### Create an Actor

Before we try to trigger the restoration, we need to find a way to validate our GCS restore
is working correctly. We will validate this by creating a detached actor. If it still
exists and functions after the head node deletion and restoration, we can confirm that the
GCS data is restored correctly.

Run following command for creating a detached actor. Please change `ha-cluster-head` to
your head node's name:

```sh

> kubectl exec -it ha-cluster-head -- python3 /home/ray/samples/detached_actor.py

Reviewer 2: Local script?

Author: This script `detached_actor.py` is defined in config map, not in local. Let me add some description here
